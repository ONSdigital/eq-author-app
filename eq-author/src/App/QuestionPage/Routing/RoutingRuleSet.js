import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { TransitionGroup } from "react-transition-group";

import routingRuleSetFragment from "graphql/fragments/routing-rule-set.graphql";

import Transition from "App/QuestionPage/Routing/Transition";
import Button from "components/Buttons/Button";
import IconText from "components/IconText";
import RoutingRuleDestinationSelector from "App/QuestionPage/Routing/RoutingRuleDestinationSelector";
import RoutingRule from "App/QuestionPage/Routing/RoutingRule";

import IconAddRule from "App/QuestionPage/Routing/icon-add-rule.svg?inline";
import { colors, radius } from "constants/theme";

const AddRuleButton = styled(Button)`
  width: 100%;
  margin-bottom: 2em;
  padding: 0.5em;
`;

const Box = styled.div`
  border: 1px solid ${colors.bordersLight};
  border-radius: ${radius};
  opacity: ${props => (props.disabled ? "0.5" : "1")};
`;

const RoutingRuleSet = ({
  ruleSet,
  onAddRoutingRule,
  onElseChange,
  ...otherProps
}) => {
  const handleAddClick = () => onAddRoutingRule(ruleSet.id);
  const handleElseChange = value =>
    onElseChange({ id: ruleSet.id, else: value });
  return (
    <React.Fragment>
      <TransitionGroup>
        {ruleSet.routingRules.map((rule, index) => (
          <Transition key={rule.id}>
            <RoutingRule
              rule={rule}
              title={index > 0 ? "Or" : null}
              key={rule.id}
              {...otherProps}
            />
          </Transition>
        ))}
      </TransitionGroup>
      <AddRuleButton
        variant="secondary"
        small
        onClick={handleAddClick}
        data-test="btn-add-rule"
      >
        <IconText icon={IconAddRule}>Add rule</IconText>
      </AddRuleButton>
      <Box>
        <RoutingRuleDestinationSelector
          id="else"
          label="ELSE"
          value={ruleSet.else}
          onChange={handleElseChange}
          data-test="select-else"
        />
      </Box>
    </React.Fragment>
  );
};

RoutingRuleSet.fragments = {
  RoutingRuleSet: routingRuleSetFragment
};

RoutingRuleSet.propTypes = {
  ruleSet: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  onAddRoutingRule: PropTypes.func.isRequired,
  onElseChange: PropTypes.func.isRequired
};

export default RoutingRuleSet;
