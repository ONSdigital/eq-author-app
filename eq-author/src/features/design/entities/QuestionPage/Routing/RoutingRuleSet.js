import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { TransitionGroup } from "react-transition-group";

import routingRuleSetFragment from "graphql/fragments/routing-rule-set.graphql";

import Transition from "./Transition";
import Button from "components/Button";
import IconText from "components/IconText";
import RoutingRuleDestinationSelector from "./RoutingRuleDestinationSelector";
import RoutingRule from "./RoutingRule";

import IconAddRule from "./icon-add-rule.svg?inline";
import { colors, radius } from "constants/theme";

const AddRuleButton = styled(Button)`
  width: 100%;
  margin-bottom: 2em;
  padding: 0.5em;
`;

const Box = styled.div`
  border: 1px solid ${colors.bordersLight};
  border-radius: ${radius};
  opacity: ${props => (props.disabled ? "0.5" : "1")};
`;

const RoutingRuleSet = ({
  ruleSet,
  destinations,
  onAddRoutingRule,
  onElseChange,
  ...otherProps
}) => {
  const handleAddClick = () => onAddRoutingRule(ruleSet.id);
  const handleElseChange = value =>
    onElseChange({ id: ruleSet.id, else: value });

  return (
    <React.Fragment>
      <TransitionGroup>
        {ruleSet.routingRules.map((rule, index) => (
          <Transition key={rule.id}>
            <RoutingRule
              rule={rule}
              title={index > 0 ? "Or" : null}
              key={rule.id}
              destinations={destinations}
              {...otherProps}
            />
          </Transition>
        ))}
      </TransitionGroup>
      <AddRuleButton
        variant="secondary"
        small
        onClick={handleAddClick}
        data-test="btn-add-rule"
      >
        <IconText icon={IconAddRule}>Add rule</IconText>
      </AddRuleButton>
      <Box>
        <RoutingRuleDestinationSelector
          id="else"
          label="ELSE"
          value={ruleSet.else}
          destinations={destinations}
          onChange={handleElseChange}
          data-test="select-else"
        />
      </Box>
    </React.Fragment>
  );
};

RoutingRuleSet.fragments = {
  RoutingRuleSet: routingRuleSetFragment
};

RoutingRuleSet.propTypes = {
  ruleSet: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  destinations: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  onAddRoutingRule: PropTypes.func.isRequired,
  onElseChange: PropTypes.func.isRequired
};

export default RoutingRuleSet;
