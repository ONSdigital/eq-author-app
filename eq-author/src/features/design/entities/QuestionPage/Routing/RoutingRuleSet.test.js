import React from "react";
import { shallow } from "enzyme";
import RoutingRuleSet from "./RoutingRuleSet";

let wrapper, props;

describe("components/RoutingRuleSet", () => {
  beforeEach(() => {
    props = {
      onAddRoutingRule: jest.fn(),
      onElseChange: jest.fn(),
      onDeleteRule: jest.fn(),
      onAddRoutingCondition: jest.fn(),
      onToggleConditionOption: jest.fn(),
      onUpdateRoutingCondition: jest.fn(),
      onUpdateConditionValue: jest.fn(),
      onThenChange: jest.fn(),
      pagesAvailableForRouting: [],
      destinations: {},
      ruleSet: {
        id: "1",
        routingRules: [
          {
            id: "2",
            operation: "And"
          },
          {
            id: "3",
            operation: "And"
          }
        ]
      }
    };

    wrapper = shallow(<RoutingRuleSet {...props} />);
  });

  it("should render children", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should allow change of ELSE condition", () => {
    const data = { name: "hello", value: "there" };
    wrapper.find("[data-test='select-else']").simulate("change", data);

    expect(props.onElseChange).toHaveBeenLastCalledWith({
      id: props.ruleSet.id,
      else: data
    });
  });

  it("should allow adding a rule", () => {
    const found = wrapper.find("[data-test='btn-add-rule']");
    found.simulate("click");
    expect(props.onAddRoutingRule).toHaveBeenCalledWith(props.ruleSet.id);
  });
});
