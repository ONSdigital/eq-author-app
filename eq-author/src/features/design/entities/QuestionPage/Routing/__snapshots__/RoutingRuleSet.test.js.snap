// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RoutingRuleSet should render children 1`] = `
<Fragment>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <Transition__RoutingComponentTransition
      key="2"
      timeout={200}
    >
      <RoutingRule
        destinations={Object {}}
        key="2"
        onAddRoutingCondition={[MockFunction]}
        onDeleteRule={[MockFunction]}
        onThenChange={[MockFunction]}
        onToggleConditionOption={[MockFunction]}
        onUpdateConditionValue={[MockFunction]}
        onUpdateRoutingCondition={[MockFunction]}
        pagesAvailableForRouting={Array []}
        rule={
          Object {
            "id": "2",
            "operation": "And",
          }
        }
        title={null}
      />
    </Transition__RoutingComponentTransition>
    <Transition__RoutingComponentTransition
      key="3"
      timeout={200}
    >
      <RoutingRule
        destinations={Object {}}
        key="3"
        onAddRoutingCondition={[MockFunction]}
        onDeleteRule={[MockFunction]}
        onThenChange={[MockFunction]}
        onToggleConditionOption={[MockFunction]}
        onUpdateConditionValue={[MockFunction]}
        onUpdateRoutingCondition={[MockFunction]}
        pagesAvailableForRouting={Array []}
        rule={
          Object {
            "id": "3",
            "operation": "And",
          }
        }
        title="Or"
      />
    </Transition__RoutingComponentTransition>
  </TransitionGroup>
  <RoutingRuleSet__AddRuleButton
    data-test="btn-add-rule"
    onClick={[Function]}
    small={true}
    variant="secondary"
  >
    <IconText
      hideText={false}
      icon={[Function]}
    >
      Add rule
    </IconText>
  </RoutingRuleSet__AddRuleButton>
  <RoutingRuleSet__Box>
    <RoutingRuleDestinationSelector
      data-test="select-else"
      destinations={Object {}}
      disabled={false}
      id="else"
      label="ELSE"
      loading={false}
      onChange={[Function]}
    />
  </RoutingRuleSet__Box>
</Fragment>
`;
